x/btcstaking/keeper/power_dist_change.go

```diff
@@ -232,6 +238,11 @@ func (k Keeper) ProcessAllPowerDistUpdateEvents(
 
 		// add this finality provider to the new cache if it has voting power
 		if fpDistInfo.TotalVotingPower > 0 {
+			// voting power is not assigned if it does not have timestamped
+			// public randomness for this height
+			if !k.FinalityKeeper.HasTimestampedPubRand(ctx, fpBTCPK, height) {
+				fpDistInfo.TotalVotingPower = 0
+			}
```

Comment:
This filtering only applies to new FPs, but not existing FPs. So if an existing active FP stops having timestamped pub rand the algorithm will keep it in the active set, which is not expected.

A proper fix could be filtering out all FPs without timestamped pub rand in `ApplyActiveFinalityProviders`. This function provides the exact definition of when a FP is "active"

Reply:
> This filtering only applies to new FPs, but not existing FPs. So if an existing active FP stops having timestamped pub rand the algorithm will keep it in the active set, which is not expected.

The filter is also applied in Line 202 for existing fps

> A proper fix could be filtering out all FPs without timestamped pub rand in ApplyActiveFinalityProviders. This function provides the exact definition of when a FP is "active". Wdyt?

I think it's a good place to put it. One minor concern is that this might introduce overhead in the sense that we need to iterate all the fp dist info and update the voting power

Reply:
> One minor concern is that this might introduce overhead in the sense that we need to iterate all the fp dist info and update the voting power

This is already the case and is inevitable. `ApplyActiveFinalityProviders` needs to find out the top N FPs in terms of voting power and the most efficient way of finding them is sorting first.

Reply:
👍  on putting it in `ApplyActiveFinalityProviders.` having this check in two separate places can lead to potential errors in the future. (and we want to refactor this area so it is better not to put landmines for our futureselves)